internal/transport/http2_client.go

```diff
@@ -1203,14 +1205,93 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		return
 	}
 
-	state := &decodeState{}
-	// Initialize isGRPC value to be !initialHeader, since if a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
-	state.data.isGRPC = !initialHeader
-	if err := state.decodeHeader(frame); err != nil {
-		t.closeStream(s, err, true, http2.ErrCodeProtocol, status.Convert(err), nil, endStream)
+	// frame.Truncated is set to true when framer detects that the current header
+	// list size hits MaxHeaderListSize limit.
+	if frame.Truncated {
+		se := status.New(codes.Internal, "peer header list size exceeded limit")
+		t.closeStream(s, se.Err(), true, http2.ErrCodeProtocol, se, nil, endStream)
 		return
 	}
 
+	// If a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
+	var mdata = make(map[string][]string)
+
+	for _, hf := range frame.Fields {
+		v, err := decodeMetadataHeader(hf.Name, hf.Value)
+		if err != nil {
+			if logger.V(logLevel) {
+				logger.Errorf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
+			}
+			mdata[hf.Name] = []string{}
+			break
+		}
+		mdata[hf.Name] = append(mdata[hf.Name], v)
```

Comment:
I'm surprised the performance is actually better with this change.  Using a `map` instead of a `struct` seems like it would be worse.

Any ideas?

Can we reduce the map accesses and make it even better?
```go
var contentType string
var grpcMessage string
for _, hf := range frame.Fields {
	....
	switch hf.Name {
	case "content-type":
		if contentType != "" {
			fail?
		}
		contentType = v
	case "grpc-message":
		if grpcMessage != "" {
			fail?
		}
		grpcMessage = v
	....
	default:
		mdata[hf.Name] = append(..., v)
	}
}
```

Reply:
I feel like my most recent push is a regression. I know we had gotten rid of some fields in favor of map-accesses but in order to reduce / remove the map accesses, we can create the necessary local fields. This has the added benefit of being a pretty clean solution (it's basically just removing the decodeState abstraction).

I did run the benchmarks again, this time for a minute instead of 30s. Here's those results:

```
unary-networkMode_Local-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      3074579      3197715     4.00%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8854.32      8892.72     0.43%
           Allocs/op       163.45       163.48     0.00%
             ReqT/op    409943.87    426362.00     4.01%
            RespT/op    409943.87    426362.00     4.01%
            50th-Lat   1.845107ms     1.8044ms    -2.21%
            90th-Lat   2.852592ms   2.674804ms    -6.23%
            99th-Lat   4.551656ms   3.953715ms   -13.14%
             Avg-Lat   1.949743ms   1.874708ms    -3.85%
           GoVersion     go1.16.2     go1.16.2
         GrpcVersion   1.38.0-dev   1.38.0-dev
```

Reply:
This is certainly not as good as last time.  I wonder if the previous run was a fluke, or maybe this time was a fluke?  There can be a lot of variance in the benchmarks.  I'll take a look at the code again and see if I can spot anything.